File: backend/models/transaction.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// TransactionType defines the type of transaction
type TransactionType string

const (
	TransactionTypeSale     TransactionType = "sale"     // Penjualan ke customer
	TransactionTypePurchase TransactionType = "purchase" // Pembelian/Setor dari customer
)

// PaymentMethod defines payment method
type PaymentMethod string

const (
	PaymentMethodCash     PaymentMethod = "cash"
	PaymentMethodTransfer PaymentMethod = "transfer"
	PaymentMethodCard     PaymentMethod = "card"
	PaymentMethodMixed    PaymentMethod = "mixed"
)

// Transaction represents a POS transaction (sale or purchase/setor)
type Transaction struct {
	ID              uint            `gorm:"primarykey" json:"id"`
	CreatedAt       time.Time       `json:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at"`
	DeletedAt       gorm.DeletedAt  `gorm:"index" json:"-"`
	TransactionCode string          `gorm:"not null;size:30" json:"transaction_code"` // unique index created manually in migration
	Type            TransactionType `gorm:"not null;size:20;index" json:"type"`
	MemberID        *uint           `gorm:"index" json:"member_id,omitempty"`
	Member          *Member         `gorm:"foreignKey:MemberID" json:"member,omitempty"`
	LocationID      uint            `gorm:"not null;index" json:"location_id"`
	Location        Location        `gorm:"foreignKey:LocationID" json:"location,omitempty"`
	CashierID       uint            `gorm:"not null;index" json:"cashier_id"`
	Cashier         User            `gorm:"foreignKey:CashierID" json:"cashier,omitempty"`

	// Financial details
	SubTotal        float64 `gorm:"not null" json:"sub_total"`
	Discount        float64 `gorm:"default:0" json:"discount"`
	DiscountPercent float64 `gorm:"default:0" json:"discount_percent"`
	Tax             float64 `gorm:"default:0" json:"tax"`
	GrandTotal      float64 `gorm:"not null" json:"grand_total"`

	// Payment
	PaymentMethod PaymentMethod `gorm:"not null;size:20" json:"payment_method"`
	PaidAmount    float64       `gorm:"not null" json:"paid_amount"`
	ChangeAmount  float64       `gorm:"default:0" json:"change_amount"`

	// Additional info
	CustomerName    string    `gorm:"size:100" json:"customer_name,omitempty"` // For non-member customers
	CustomerPhone   string    `gorm:"size:20" json:"customer_phone,omitempty"`
	Notes           string    `gorm:"size:500" json:"notes"`
	Status          string    `gorm:"not null;size:20;default:'completed'" json:"status"` // completed, cancelled, refunded
	TransactionDate time.Time `gorm:"not null;index" json:"transaction_date"`

	// Relations
	Items []TransactionItem `gorm:"foreignKey:TransactionID" json:"items,omitempty"`
}

// TransactionItem represents items in a transaction
type TransactionItem struct {
	ID            uint           `gorm:"primarykey" json:"id"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"-"`
	TransactionID uint           `gorm:"not null;index" json:"transaction_id"`
	Transaction   Transaction    `gorm:"foreignKey:TransactionID" json:"transaction,omitempty"`

	// For sale transactions - existing stock
	StockID *uint  `gorm:"index" json:"stock_id,omitempty"`
	Stock   *Stock `gorm:"foreignKey:StockID" json:"stock,omitempty"`

	// For purchase/setor transactions - no existing stock, manual entry
	ProductID      *uint         `gorm:"index" json:"product_id,omitempty"`
	Product        *Product      `gorm:"foreignKey:ProductID" json:"product,omitempty"`
	GoldCategoryID *uint         `gorm:"index" json:"gold_category_id,omitempty"`
	GoldCategory   *GoldCategory `gorm:"foreignKey:GoldCategoryID" json:"gold_category,omitempty"`

	// Item details
	ItemName     string  `gorm:"not null;size:100" json:"item_name"`
	Barcode      string  `gorm:"size:50" json:"barcode"`
	Weight       float64 `gorm:"not null" json:"weight"` // Berat dalam gram
	PricePerGram float64 `gorm:"not null" json:"price_per_gram"`
	UnitPrice    float64 `gorm:"not null" json:"unit_price"` // Total price for this item
	Quantity     int     `gorm:"not null;default:1" json:"quantity"`
	Discount     float64 `gorm:"default:0" json:"discount"`
	SubTotal     float64 `gorm:"not null" json:"sub_total"`
	Notes        string  `gorm:"size:255" json:"notes"`
}

// PurchaseItem represents items bought from customers (setor)
type PurchaseItem struct {
	ID             uint           `gorm:"primarykey" json:"id"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
	TransactionID  uint           `gorm:"not null;index" json:"transaction_id"`
	GoldCategoryID uint           `gorm:"not null;index" json:"gold_category_id"`
	GoldCategory   GoldCategory   `gorm:"foreignKey:GoldCategoryID" json:"gold_category,omitempty"`
	Weight         float64        `gorm:"not null" json:"weight"`         // Berat dalam gram
	PricePerGram   float64        `gorm:"not null" json:"price_per_gram"` // Harga beli per gram
	TotalPrice     float64        `gorm:"not null" json:"total_price"`
	Condition      string         `gorm:"size:50" json:"condition"` // Kondisi barang
	Notes          string         `gorm:"size:255" json:"notes"`
}

File: mobile/lib/models/transaction.dart
import 'package:equatable/equatable.dart';
import 'member.dart';
import 'location.dart';
import 'user.dart';
import 'transaction_item.dart';

class Transaction extends Equatable {
  final int id;
  final String transactionCode;
  final String type; // sale, purchase
  final int? memberId;
  final Member? member;
  final int locationId;
  final Location? location;
  final int cashierId;
  final User? cashier;
  
  // Financial details
  final double subTotal;
  final double discount;
  final double discountPercent;
  final double tax;
  final double grandTotal;
  
  // Payment
  final String paymentMethod; // cash, transfer, card, mixed
  final double paidAmount;
  final double changeAmount;
  
  // Additional info
  final String? customerName;
  final String? customerPhone;
  final String notes;
  final String status; // completed, cancelled, refunded
  final DateTime transactionDate;
  
  final List<TransactionItem>? items;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const Transaction({
    required this.id,
    required this.transactionCode,
    required this.type,
    this.memberId,
    this.member,
    required this.locationId,
    this.location,
    required this.cashierId,
    this.cashier,
    required this.subTotal,
    this.discount = 0,
    this.discountPercent = 0,
    this.tax = 0,
    required this.grandTotal,
    required this.paymentMethod,
    required this.paidAmount,
    this.changeAmount = 0,
    this.customerName,
    this.customerPhone,
    this.notes = '',
    this.status = 'completed',
    required this.transactionDate,
    this.items,
    this.createdAt,
    this.updatedAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'] ?? 0,
      transactionCode: json['transaction_code'] ?? '',
      type: json['type'] ?? 'sale',
      memberId: json['member_id'],
      member: json['member'] != null ? Member.fromJson(json['member']) : null,
      locationId: json['location_id'] ?? 0,
      location: json['location'] != null ? Location.fromJson(json['location']) : null,
      cashierId: json['cashier_id'] ?? 0,
      cashier: json['cashier'] != null ? User.fromJson(json['cashier']) : null,
      subTotal: (json['sub_total'] ?? 0).toDouble(),
      discount: (json['discount'] ?? 0).toDouble(),
      discountPercent: (json['discount_percent'] ?? 0).toDouble(),
      tax: (json['tax'] ?? 0).toDouble(),
      grandTotal: (json['grand_total'] ?? 0).toDouble(),
      paymentMethod: json['payment_method'] ?? 'cash',
      paidAmount: (json['paid_amount'] ?? 0).toDouble(),
      changeAmount: (json['change_amount'] ?? 0).toDouble(),
      customerName: json['customer_name'],
      customerPhone: json['customer_phone'],
      notes: json['notes'] ?? '',
      status: json['status'] ?? 'completed',
      transactionDate: json['transaction_date'] != null
          ? DateTime.parse(json['transaction_date'])
          : DateTime.now(),
      items: json['items'] != null
          ? (json['items'] as List).map((i) => TransactionItem.fromJson(i)).toList()
          : null,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : null,
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'transaction_code': transactionCode,
      'type': type,
      'member_id': memberId,
      'location_id': locationId,
      'cashier_id': cashierId,
      'sub_total': subTotal,
      'discount': discount,
      'discount_percent': discountPercent,
      'tax': tax,
      'grand_total': grandTotal,
      'payment_method': paymentMethod,
      'paid_amount': paidAmount,
      'change_amount': changeAmount,
      'customer_name': customerName,
      'customer_phone': customerPhone,
      'notes': notes,
      'status': status,
      'transaction_date': transactionDate.toIso8601String(),
      if (items != null) 'items': items!.map((i) => i.toJson()).toList(),
    };
  }

  Transaction copyWith({
    int? id,
    String? transactionCode,
    String? type,
    int? memberId,
    Member? member,
    int? locationId,
    Location? location,
    int? cashierId,
    User? cashier,
    double? subTotal,
    double? discount,
    double? discountPercent,
    double? tax,
    double? grandTotal,
    String? paymentMethod,
    double? paidAmount,
    double? changeAmount,
    String? customerName,
    String? customerPhone,
    String? notes,
    String? status,
    DateTime? transactionDate,
    List<TransactionItem>? items,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Transaction(
      id: id ?? this.id,
      transactionCode: transactionCode ?? this.transactionCode,
      type: type ?? this.type,
      memberId: memberId ?? this.memberId,
      member: member ?? this.member,
      locationId: locationId ?? this.locationId,
      location: location ?? this.location,
      cashierId: cashierId ?? this.cashierId,
      cashier: cashier ?? this.cashier,
      subTotal: subTotal ?? this.subTotal,
      discount: discount ?? this.discount,
      discountPercent: discountPercent ?? this.discountPercent,
      tax: tax ?? this.tax,
      grandTotal: grandTotal ?? this.grandTotal,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paidAmount: paidAmount ?? this.paidAmount,
      changeAmount: changeAmount ?? this.changeAmount,
      customerName: customerName ?? this.customerName,
      customerPhone: customerPhone ?? this.customerPhone,
      notes: notes ?? this.notes,
      status: status ?? this.status,
      transactionDate: transactionDate ?? this.transactionDate,
      items: items ?? this.items,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [id, transactionCode, type, grandTotal, status];
}
